sematic/plugins/abstract_storage.py

```diff
@@ -0,0 +1,57 @@
+# Standard Library
+import abc
+import enum
+from dataclasses import dataclass
+from typing import Any
+
+
+class PayloadType(enum.Enum):
+    URL = "URL"
+    BYTES = "BYTES"
+
+
+@dataclass
+class ReadPayload:
+    type_: PayloadType
+    content: Any
+
+
+class AbstractStorage(abc.ABC):
+    """
+    Abstract base class to represent a key/value storage engine.
+    """
+
+    @abc.abstractmethod
+    def set(self, key: str, value: bytes):
+        """
+        Sets value for key.
+
+        This is a resolver-side API. Key will be provided by the server and will
+        be the output of get_write_location.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_write_location(self, namespace: str, key: str) -> str:
+        """
+        Gets write location for namespace/key.
+
+        This is a server-side API. It is used to return write locations to the resolver.
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_read_payload(self, namespace: str, key: str) -> ReadPayload:
+        """
+        Get a read payload for namespace/key.
+
+        This is a server-side API. It is used to return data to the resolver.
+        The returned payload can be a URL to redirect to (PayloadType.URL) or a
+        binary content to return as is to the resolver (PayloadType.BYTES).
+        """
+        pass
+
```

Comment:
@augray @chance-an note the change of API for `AbstractStorage`. This is to implement suggestions by you two.

The `AbstractStorage.get` API was removed as the bytes payload will be returned by the server, either from a direct disk read, or from a redirect to a signed URL as suggested by @chance-an.

The server endpoint will get a `ReadPayload` object from the storage engine. Depending on `read_payload.type_`, which is of type `PayloadType`, the endpoint will return `read_payload.content` (`PayloadType.BYTES`) or redirect (`PayloadType.URL`).